### PR TITLE
Hello Conversion Layer Adapter

### PR DESCRIPTION
**Pull request by ueman**
Working towards https://github.com/cfug/dio/issues/1740

After this is published, we can remove the code from the `native_dio_adapter` and depend on this package instead.

## TODO

- [x] Tests
- [x] CI

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

**Comment by SpeedReach** [comment]
So this fixes #1835 since we are replacing native adapter with this instead?

**Comment by zhangyaobin** [comment]
hi, i have used ConversionLayerAdapter to solve “Web Environment cannot obtain real-time stream”,but it doesn't work, it still return the result by one time, here is my example, please help me figure out what i am wrong:
Map<String, dynamic> bodyWrap = {'params': body};
  RequestOptions options =
      RequestOptions(method: "post",contentType: "application/json",responseType: ResponseType.stream);
  options.headers = ChatHttpUtil.getInstance().getDefaultHead();
  options.data = json.encode(bodyWrap);
  options.baseUrl = "http://test-inside-city-ai-assistant.songguo7.com/";
  options.path = "city-ai-assistant/backend/chat/sse/send/v2";
  final cla = ConversionLayerAdapter(BrowserClient());
  final response = await cla.fetch(
    options,
    null,
    null,
  );

  if (response.statusCode == 200 || response.statusCode == 201) {
    ChatData chatData = ChatData(false, [], "", 0);
    ctx.state.chatDataList.add(chatData);
    Stream<Uint8List> stream = response.stream;
    stream.listen((event) {
      print("receiver event");
      Uint8List uint8list = event;
      String string = utf8.decode(uint8list);
      List<String> items = string.split("data:");
      print("List<String> items=$items");
      items.forEach((element) {
        element = element.trim();
        if (element.startsWith("{") && element.endsWith("}")) {
          Map<String, dynamic> map = jsonDecode(element);
          element = element.replaceAll("\n", '\\n');
          ChatBaseInfo chatBaseInfo = ChatBaseInfo.fromJson(map);
          if (chatBaseInfo.code == 200) {
            ctx.state.chatDataList.last.messageId =
                chatBaseInfo.data!.messageId;
            ctx.state.sessionId = chatBaseInfo.data!.sessionId!;
            handleData(
                chatBaseInfo.data!.content!, ctx.state.chatDataList.last);
            if ("stop" == chatBaseInfo.data!.finishReason) {
              ctx.state.isSendingMessage = false;
            }
          } else {
            ctx.state.chatDataList.last.infos
                .add(ChatInfo(1, text: chatBaseInfo.message));
            ctx.state.isSendingMessage = false;
          }
          ctx.dispatch(ChatActionCreator.onUpdate());
        }
      });
    });

**Comment by zhangyaobin** [comment]
here is ConversionLayerAdapter, i just add this "request.body=options.data":
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:dio/dio.dart';
import 'package:http/http.dart';

/// A conversion layer which translates Dio HTTP requests to
/// [http](https://pub.dev/packages/http) compatible requests.
/// This way there's no need to implement custom [HttpClientAdapter]
/// for each platform. Therefore, the required effort to add tests is kept
/// to a minimum. Since `CupertinoClient` and `CronetClient` depend anyway on
/// `http` this also doesn't add any additional dependency.
class ConversionLayerAdapter implements HttpClientAdapter {
  final Client client;

  ConversionLayerAdapter(this.client);

  @override
  Future<ResponseBody> fetch(
      RequestOptions options,
      Stream<Uint8List>? requestStream,
      Future<dynamic>? cancelFuture,
      ) async {
    print("ConversionLayerAdapter fetch");
    final request = await _fromOptionsAndStream(options, requestStream);
    final response = await client.send(request);
    return response.toDioResponseBody();
  }

  @override
  void close({bool force = false}) => client.close();

  Future<BaseRequest> _fromOptionsAndStream(
      RequestOptions options,
      Stream<Uint8List>? requestStream,
      ) async {
    final request = Request(
      options.method,
      options.uri,
    );

    // request.bodyBytes=json.encode(options.data).codeUnits;
   if(options.data!=null){
      request.body=options.data;
    }

    // print(" request.body=${ request.body}");
    request.headers.addAll(
      Map.fromEntries(
        options.headers.entries.map((e) => MapEntry(e.key, e.value.toString())),
      ),
    );

    request.followRedirects = options.followRedirects;
    request.maxRedirects = options.maxRedirects;

    if (requestStream != null) {
      final completer = Completer<Uint8List>();
      final sink = ByteConversionSink.withCallback(
            (bytes) => completer.complete(Uint8List.fromList(bytes)),
      );
      requestStream.listen(
        sink.add,
        onError: completer.completeError,
        onDone: sink.close,
        cancelOnError: true,
      );
      final bytes = await completer.future;
      request.bodyBytes = bytes;
    }
    return request;
  }
}

extension on StreamedResponse {
  ResponseBody toDioResponseBody() {
    final dioHeaders = headers.entries.map((e) => MapEntry(e.key, [e.value]));
    return ResponseBody(
      stream.cast<Uint8List>(),
      statusCode,
      headers: Map.fromEntries(dioHeaders),
      isRedirect: isRedirect,
      statusMessage: reasonPhrase,
    );
  }
}

**Comment by AlexV525** [comment]
@ueman Is this still on your track?

**Comment by AlexV525** [comment]
I've also requested license and authoring info changes. FYI @ueman

**Comment by AlexV525** [comment]
@86b7a7b indicates that we're not using the `StreamedRequest` for the `native_dio_adapter`, which will cost bytes amount of memory waste when sending with the request stream. It should be addressed separately before we roll the layer for the adapter.

File: plugins/http_compatibility_layer/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+### Install
+
+Add the `dio_http_compatibility_layer` package to your
+[pubspec dependencies](https://pub.dev/packages/dio_http_compatibility_layer/install).
+
+### Example
+
+```dart
+Dio().httpClientAdapter = ConversionLayerAdapter(Client());
+```
```
Comment (ueman) [inline plugins/http_compatibility_layer/README.md:24]:
```suggestion
import 'package:dio/dio.dart';
import 'package:dio_http_compatibility_layer/dio_http_compatibility_layer.dart';
import 'package:http/http.dart';

void main() {
  // Start in the `http` world. You can use `http`, `cronet_http`, `cupertino_http` and other http compatible packages.
  final httpClient = Client();

  // Make the httpClient compatible via the ConversionLayerAdapter class
  final dioAdapter = ConversionLayerAdapter(httpClient);

  // Make dio use the httpClient via the conversion layer
  final dio = Dio()..httpClientAdapter = dioAdapter;

  // Make a request
  final response = await dio.get('https://dart.dev');
  print(response);
}
```

**Comment by AlexV525** [comment]
> I think we can still release it without converting `native_dio_adapter` and deal with the `StreamedRequest` problem later.

Yeah I'm not expecting to convert it in recent weeks.

**Comment by AlexV525** [comment]
I've removed the `http` from its name since we might add more compatibilities to the packages in the future.

File: plugins/compatibility_layer/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+[![popularity](https://img.shields.io/pub/popularity/dio_compatibility_layer)](https://pub.dev/packages/dio_compatibility_layer/score)
+[![pub points](https://img.shields.io/pub/points/dio_compatibility_layer)](https://pub.dev/packages/dio_compatibility_layer/score)
+
+If you encounter bugs, consider fixing it by opening a PR or at least contribute a failing test case.
+
+This is adapters for [Dio](https://pub.dev/packages/dio)
+which enables you to make use of other HTTP clients as underlying implementation.
+
+Supported client base currently:
+- `http`
```
Comment (ueman) [inline plugins/compatibility_layer/README.md:14]:
```suggestion
This package contains adapters for [Dio](https://pub.dev/packages/dio)
which enables you to make use of other HTTP clients as the underlying implementation.

Currently, it supports compatibility with
- [`http`](https://pub.dev/packages/http)
```

**Review by ueman — COMMENTED** [review]
Thanks for bringing this over the finish line

**Comment by AlexV525** [comment]
I'll check with @CaiJingLong to see if we need extra support around publishing tools for the new package.

**Pull request merged**